### PR TITLE
SALTO-2066: make dump internally sync

### PR DESCRIPTION
**Pull request by roironn**
_Make dump internally sync by adding promise value watchers_

---

The dump method is currently async since the functions dump method is async.
Tests showed that removing this can improve the dump performance in about 50%.

This is can be done by creating value promise watchers and awaiting them all at once like we do in the parser. 

---
_Release Notes_: 
_Improved the NaCL dump methods performance_

---
_User Notifications_: 
_NA_

### CODE REVIEW

**Comment by ori-moisis** [comment]
why not make the functions interface sync?
dump doesn't really have to wait for the static file source to finished writing the file (we could manage the "pending" writes in the static file source and add a "waitPendingWrites" to it maybe), the part that creates the function expression is sync anyway...

I think we can also do that for parse (and maybe gain some performance there as well), but probably in a separate effort, doesn't have to be now

File: packages/workspace/src/parser/dump.ts [hunk 1]
```diff
@@ -164,11 +180,15 @@ const wrapBlocks = (blocks: DumpedHclBlock[]): DumpedHclBlock => ({
 
 export const dumpElements = async (
   elements: Readonly<Element>[], functions: Functions = {}, indentationLevel = 0
-): Promise<string> =>
-  hclDump(
-    wrapBlocks(await awu(elements).map(e => dumpElementBlock(e, functions)).toArray()),
-    indentationLevel
+): Promise<string> => {
+  const valuePromiseWatchers: ValuePromiseWatcher[] = []
```
Comment (dorsev) [inline packages/workspace/src/parser/dump.ts:184]: Some questions:
What is ValuePromiseWatcher?
Also, what does internally sync mean?

Comment (roironn) [reply packages/workspace/src/parser/dump.ts:184]: 1. Internally sync - removed asyncness from all of the internal functions used by the code (note, all of the functions that are no longer async)
2. The methods were async sync the dump method for functions is async. In order to achieve this we add promises as values during the DFS, and keep a context of where the promises were added in the `valuePromiseWatchers` array. When the DFS is done, we await the promises and replace the values.

Comment (dorsev) [reply packages/workspace/src/parser/dump.ts:184]: makes sense, Thanks for the explanation

**Comment by roironn** [comment]
> why not make the functions interface sync? dump doesn't really have to wait for the static file source to finished writing the file (we could manage the "pending" writes in the static file source and add a "waitPendingWrites" to it maybe), the part that creates the function expression is sync anyway...
> 
> I think we can also do that for parse (and maybe gain some performance there as well), but probably in a separate effort, doesn't have to be now

I agree that we can do some ( a lot ) of changes for the static file interface - but that would result in a much bigger code change (which might also effect other reops). The purpose of this PR it to utilize existing mechanisms to make this code quicker without modifying any external interfaces.

File: packages/workspace/src/parser/dump.ts [hunk 2]
```diff
@@ -75,18 +75,31 @@ const dumpAttributes = async (value: Value, functions: Functions): Promise<Value
...
   }
-
   if (_.isArray(value)) {
-    return Promise.all(value.map(x => dumpAttributes(x, functions)))
+    const retArray: Value[] = []
+    value.forEach((x, key) => {
+      retArray.push(dumpAttributes(x, functions, valuePromiseWatchers))
+      addValuePromiseWatcher(valuePromiseWatchers, retArray, key)
+    })
+    return retArray
```
Comment (ori-moisis) [inline packages/workspace/src/parser/dump.ts:84]: creating an array with `forEach` and `push` seems to be significantly slower than using `map`

tested with dummy code:
```javascript
const base = new Array(10000000).fill(1)

const time = (name, func) => {
  const start = Date.now()
  func()
  const end = Date.now()
  console.log('%s took %d', name, end - start)
}

const withForEach = () => {
  const context = []
  const arr = []
  base.forEach((item, idx) => {
    arr.push(item + idx)
    if (arr[idx] < 10) {
      context.push(idx)
    }
  })
}

const withMap = () => {
  const context = []
  const arr = base.map((item, idx) => {
    const res = item + idx
    if (res < 10) {
      context.push(idx)
    }
    return res
  })
}

time('forEach', withForEach)
time('map', withMap)
```
```bash
node bla.js
forEach took 349
map took 169
```
(results were fairly consistent, switching the order didn't seem to make a difference)

File: packages/workspace/src/parser/dump.ts [hunk 3]
```diff
@@ -75,18 +75,31 @@ const dumpAttributes = async (value: Value, functions: Functions): Promise<Value
...
   }
 
-  return mapValuesAsync(value, async val => dumpAttributes(val, functions))
+  const retObj: Record<string, Value> = {}
+  Object.entries(value).forEach(([key, val]) => {
+    retObj[key] = dumpAttributes(val, functions, valuePromiseWatchers)
+    addValuePromiseWatcher(valuePromiseWatchers, retObj, key)
+  })
+  return retObj
```
Comment (ori-moisis) [inline packages/workspace/src/parser/dump.ts:92]: maybe there is a similar potential gain with using `map` and `Object.fromEntries`? not sure about that but might be worth checking

**Comment by ori-moisis** [comment]
btw, might be worth checking very large inputs to see what happens with memory consumption now.
maybe we need to switch the arrays to iterators to streamline memory access a little?
(raising this now since the removal of async parts would change the flow of memory access, might make it faster for small inputs and slower / crash for very large inputs)

File: packages/workspace/src/parser/dump.ts (hunk 2 above)
Comment (roironn) [reply packages/workspace/src/parser/dump.ts:84]: Yes - `map` is faster - but we need the pointer to to reValue in order to save a promise watcher for it before we start the loop.

I can check if running `map` on value, preallocating `retValue` and using index insertion will be quicker.

File: packages/workspace/src/parser/dump.ts (hunk 3 above)
Comment (roironn) [reply packages/workspace/src/parser/dump.ts:92]: Same issue as above - we need a pointer to `retObj`

File: packages/workspace/src/parser/dump.ts (hunk 2 above)
Comment (ori-moisis) [reply packages/workspace/src/parser/dump.ts:84]: two options
1. iterate twice (once to create the map, then a second time to find the promises)
2. single iteration where you just remember the index and then add the "watcher" thing after the `map` is done

Comment (roironn) [reply packages/workspace/src/parser/dump.ts:84]: Won't the push to the promise indices get us to the same spot in which we started?

Comment (roironn) [reply packages/workspace/src/parser/dump.ts:84]: Either way - I can tinker with it a bit and choose what ever is fastest :)

**Pull request merged**